syntax = "proto3";
package protos.common;
option java_package = "es.bsc.dataclay.communication.grpc.messages.common";

message Credential {
	string password = 1;
}

message ParamOrReturn {
	string objectID = 1;
	bytes serializedParam = 2;
}

message EmptyMessage {

}

enum Langs {
	LANG_NONE = 0;
	LANG_JAVA = 1;
	LANG_PYTHON = 2;
}

message SerializedParametersOrReturn { 
	int32 numParams = 1;
	map<int32, ImmutableParamOrReturn> immParams = 2;
	map<int32, LanguageParamOrReturn> langParams = 3;
	map<int32, ObjectWithDataParamOrReturn> volatileParams = 4;
	map<int32, PersistentParamOrReturn> persParams = 5;
}

message ImmutableParamOrReturn {
	bytes objbytes = 1;
}

message PersistentParamOrReturn {
	string oid = 1;
	string hint = 2;
	string classID = 3;
	string extDataClayID = 4;
}

message ObjectWithDataParamOrReturn {
	string oid = 1;
	string classid = 2;
	DataClayObjectMetaData metadata = 3;
	bytes objbytes = 4;
}

message LanguageParamOrReturn { 
	DataClayObjectMetaData metadata = 1;
	bytes objbytes = 2;
}

message DataClayObjectMetaData { 
	map<int32, string> oids = 1;
	map<int32, string> classids = 2;
	map<int32, string> hints = 3;
	int32 numRefs = 4;
	string origObjectID = 5;
	string rootLocation = 6;
	string originLocation = 7;
	repeated string replicaLocations = 8;
	string alias = 9;
	bool isReadOnly = 10;
}

message PersistentObjectInDB {
	DataClayObjectMetaData metadata = 1;
	bytes data = 2;
}

message RegistrationInfo {
	string objectID = 1;
	string classID = 2;
	string sessionID = 3;
	string dataSetID = 4;
	string alias = 5;
}

message FederatedObjectInfo {
	string objectID = 1;
	string className = 2;
	string nameSpace = 3;
	repeated string alias = 4;
}

message GetTracesResponse { 
	map<string, bytes> traces = 1;
	protos.common.ExceptionInfo excInfo = 2;
	
}

message ExceptionInfo {
	bool isException = 1;
	bytes serializedException = 2;
	bytes exceptionMessage = 3;
}

message MetaDataInfo {
	string objectID = 1;
	bool isReadOnly = 2;
	string datasetID = 3;
	string metaclassID = 4;
	repeated string locations = 5;
	string alias = 6;
	string ownerID = 7;
}

message StorageLocationInfo {
	string id = 1;
	string hostname = 2;
	string name = 3;
	int32 port = 4;
}

message ExecutionEnvironmentInfo {
	string id = 1;
	string hostname = 2;
	string name = 3;
	int32 port = 4;
	Langs language = 5;
	string dataClayInstanceID = 6;
}

message DataClayInstance {
	string id = 1;
	repeated string hosts = 2;
	repeated int32 ports = 3;
}


message GetNumObjectsResponse {
	int32 numObjs = 1;
	protos.common.ExceptionInfo excInfo = 2;
}

// New

message ExecutionEnvironment {
	string id = 1;
	string hostname = 2;
	string name = 3;
	int32 port = 4;
	Langs language = 5;
	string dataclay_id = 6;
}